File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.cpp
class Solution {
public:
    long long dp[100001]={[0 ... 100000]=-1};
    long long recur(int idx,vector<vector<int>>&q){
        if(idx>=q.size())return 0;
        if(dp[idx]!=-1)return dp[idx];
        
        return dp[idx]= max(q[idx][0]+recur(idx+q[idx][1]+1,q),recur(idx+1,q));
        
    }
    
    long long mostPoints(vector<vector<int>>& questions) {
        return recur(0,questions);
    }
};

File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.java
class Solution {
    long []dp;
    public long memo(int idx,int[][] q){
        if(idx>=q.length)return 0;
        if(dp[idx]!=-1)return dp[idx];
        return dp[idx]= Math.max(q[idx][0]+memo(idx+1+q[idx][1],q),memo(idx+1,q));
    }
    public long mostPoints(int[][] questions) {
        int n=questions.length;
        dp=new long[n+1];
        // Arrays.fill(dp,-1);
        // return memo(0,questions);
        for(int i=n-1;i>=0;i--){
            dp[i]=Math.max(questions[i][0]+dp[(i+questions[i][1]+1)<n?(i+questions[i][1]+1):n],dp[i+1]);
        }
        
        return dp[0];

        }
}